rockin/plugins/drilling_machine/drilling_machine_simulation_thread.cpp

```diff
@@ -0,0 +1,142 @@
+    clips->add_function("drilling-machine-move-drill-down", sigc::slot<void>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_move_drill_down)));
+    clips->add_function("drilling-machine-get-state", sigc::slot<int>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_get_device_state)));
+    clips->add_function("drilling-machine-is-device-connected", sigc::slot<int>(sigc::mem_fun(*this, &DrillingMachineSimulationThread::clips_is_device_connected)));
+
+    if (!clips->build("(deffacts have-feature-drilling-machine (have-feature DrillingMachine))"))
+        logger->log_warn("DrillingMachineSim", "Failed to build deffacts have-feature-drilling-machine");
+}
+
+void DrillingMachineSimulationThread::finalize()
+{
+}
+
+void DrillingMachineSimulationThread::loop()
+{
+    boost::this_thread::sleep(boost::posix_time::milliseconds(cfg_timer_interval_));
+    fawkes::MutexLocker lock(clips_mutex);
+    if (last_command_msg_.command() == DrillingMachineCommand::MOVE_DOWN)
+    {
+        if (tick_ > 1)
+        {
+            last_status_msg_.set_state(DrillingMachineStatus::MOVING_DOWN);
+            tick_ -= 1;
+        } else
+        {
+            last_status_msg_.set_state(DrillingMachineStatus::AT_BOTTOM);
+        }
+    } else if (last_command_msg_.command() == DrillingMachineCommand::MOVE_UP)
+    {
+        if (tick_ < ENCODER_MAX_)
+        {
+            last_status_msg_.set_state(DrillingMachineStatus::MOVING_UP);
+            tick_ += 1;
+        } else
+        {
+            last_status_msg_.set_state(DrillingMachineStatus::AT_TOP);
+        }
+    }
+}
+
+DrillingMachineStatus::State DrillingMachineSimulationThread::clips_get_device_state()
+{
+    fawkes::MutexLocker lock(clips_mutex);
+
+    if (last_status_msg_.has_state())
+    {
+        return last_status_msg_.state();
+    }
+    return DrillingMachineStatus::UNKNOWN;
+}
+
+int DrillingMachineSimulationThread::clips_is_device_connected()
+{
+    fawkes::MutexLocker lock(clips_mutex);
+    return (last_status_msg_.has_is_device_connected() && last_status_msg_.is_device_connected());
+}
+
+void DrillingMachineSimulationThread::clips_move_drill_up()
+{
+    moveDrill(DrillingMachineCommand::MOVE_UP);
+}
+
+void DrillingMachineSimulationThread::clips_move_drill_down()
+{
+    moveDrill(DrillingMachineCommand::MOVE_DOWN);
+}
+
+void DrillingMachineSimulationThread::moveDrill(DrillingMachineCommand::Command drill_command)
+{
+    // prevent sending to many messages to the device
+    boost::posix_time::time_duration time_diff;
+    time_diff = boost::posix_time::microsec_clock::local_time() - last_sent_command_timestamp_;
+    if (time_diff.total_milliseconds() < 200)
+        return;
+
+    fawkes::MutexLocker lock(clips_mutex);
+    last_command_msg_.set_command(drill_command);
+    last_sent_command_timestamp_ = boost::posix_time::microsec_clock::local_time();
+
+    logger->log_info("DrillingMachineSim", "Send drill command: %d", drill_command);
```

Comment:
should I modify that in the other plugins as well? And the non-simulated versions?